core/c_api.cpp

```diff
@@ -0,0 +1,492 @@
+                                       gko_executor other_exec_st_ptr)
+{
+    return exec_st_ptr->shared_ptr->memory_accessible(
+        other_exec_st_ptr->shared_ptr);
+}
+
+void ginkgo_executor_synchronize(gko_executor exec_st_ptr)
+{
+    exec_st_ptr->shared_ptr->synchronize();
+}
+
+//---------------------------- CPU -----------------------------
+gko_executor ginkgo_executor_omp_create()
+{
+    return new gko_executor_st{gko::OmpExecutor::create()};
+}
+
+gko_executor ginkgo_executor_reference_create()
+{
+    return new gko_executor_st{gko::ReferenceExecutor::create()};
+}
+
+size_t ginkgo_executor_cpu_get_num_cores(gko_executor exec_st_ptr)
+{
+    if (auto omp_exec = std::dynamic_pointer_cast<gko::OmpExecutor>(
+            exec_st_ptr->shared_ptr)) {
+        return omp_exec->get_num_cores();
+    } else if (auto ref_exec =
+                   std::dynamic_pointer_cast<gko::ReferenceExecutor>(
+                       exec_st_ptr->shared_ptr)) {
+        return ref_exec->get_num_cores();
+    } else {
+        return 0;
+    }
+}
+
+size_t ginkgo_executor_cpu_get_num_threads_per_core(gko_executor exec_st_ptr)
+{
+    if (auto omp_exec = std::dynamic_pointer_cast<gko::OmpExecutor>(
+            exec_st_ptr->shared_ptr)) {
+        return omp_exec->get_num_threads_per_core();
+    } else if (auto ref_exec =
+                   std::dynamic_pointer_cast<gko::ReferenceExecutor>(
+                       exec_st_ptr->shared_ptr)) {
+        return ref_exec->get_num_threads_per_core();
+    } else {
+        return 0;
+    }
+}
+
+//---------------------------- GPU -----------------------------
+size_t ginkgo_executor_gpu_get_device_id(gko_executor exec_st_ptr)
+{
+    if (auto cuda_exec = std::dynamic_pointer_cast<gko::CudaExecutor>(
+            exec_st_ptr->shared_ptr)) {
+        return cuda_exec->get_device_id();
+    } else if (auto hip_exec = std::dynamic_pointer_cast<gko::HipExecutor>(
+                   exec_st_ptr->shared_ptr)) {
+        return hip_exec->get_device_id();
+    } else if (auto dpcpp_exec = std::dynamic_pointer_cast<gko::DpcppExecutor>(
+                   exec_st_ptr->shared_ptr)) {
+        return dpcpp_exec->get_device_id();
+    } else {
+        return 0;
+    }
+}
+
+// CUDA/HIP
+gko_executor ginkgo_executor_cuda_create(size_t device_id,
+                                         gko_executor exec_st_ptr)
+{
+    if (auto omp_exec = std::dynamic_pointer_cast<gko::OmpExecutor>(
+            exec_st_ptr->shared_ptr)) {
+        return new gko_executor_st{
+            gko::CudaExecutor::create(device_id, exec_st_ptr->shared_ptr)};
+    } else if (auto reference_exec =
+                   std::dynamic_pointer_cast<gko::ReferenceExecutor>(
+                       exec_st_ptr->shared_ptr)) {
+        return new gko_executor_st{
+            gko::CudaExecutor::create(device_id, exec_st_ptr->shared_ptr)};
+    } else {
+        return nullptr;
+    }
+}
+
+size_t ginkgo_executor_cuda_get_num_devices()
```

Comment:
this should return an `int` instead of `size_t`, since that's what the ginkgo function returns.
Or is the `size_t` required from the julia side.
This also applies to many of the executor functions.